File: scripts/verify.ts
/**
 * Sanity harness: build every part from a JSON file and print vertex/face
 * counts, bounding box and a coordinate checksum, in the same format as the
 * companion verify.py — so `diff` tells us whether the TS port matches the
 * Python original.
 */
import { readFileSync } from 'node:fs';
import {
  buildBaseplateFromAny, buildInsertsFromAny, buildShellFromAny,
  buildWallsFromAny, isAssembly, resolveKeylist,
} from '../src/core/core';
import type { Entry, Mesh } from '../src/core/types';

function report(label: string, mesh: Mesh) {
  const { vertices, faces } = mesh;
  let minX = Infinity, minY = Infinity, minZ = Infinity;
  let maxX = -Infinity, maxY = -Infinity, maxZ = -Infinity;
  let sum = 0;
  for (const [x, y, z] of vertices) {
    minX = Math.min(minX, x); maxX = Math.max(maxX, x);
    minY = Math.min(minY, y); maxY = Math.max(maxY, y);
    minZ = Math.min(minZ, z); maxZ = Math.max(maxZ, z);
    sum += x + 2 * y + 3 * z;
  }
  const nTris = faces.reduce((s, f) => s + Math.max(0, f.length - 2), 0);
  // Manifold check on polygon edges: every edge shared by exactly 2 faces.
  const edgeCount = new Map<string, number>();
  for (const f of faces) {
    for (let i = 0; i < f.length; i++) {
      const a = f[i], b = f[(i + 1) % f.length];
      const k = a < b ? `${a},${b}` : `${b},${a}`;
      edgeCount.set(k, (edgeCount.get(k) ?? 0) + 1);
    }
  }
  let bad = 0;
  for (const c of edgeCount.values()) if (c !== 2) bad++;
  const fmt = (v: number) => v.toFixed(4);
  console.log(
    `${label}: verts=${vertices.length} faces=${faces.length} tris=${nTris} ` +
    `bbox=[${fmt(minX)},${fmt(minY)},${fmt(minZ)}]..[${fmt(maxX)},${fmt(maxY)},${fmt(maxZ)}] ` +
    `checksum=${sum.toFixed(3)} nonmanifold_edges=${bad}`,
  );
}

const path = process.argv[2];
const raw = JSON.parse(readFileSync(path, 'utf8'));
const entries: Entry[] = Array.isArray(raw) ? raw : [raw];

for (const entry of entries) {
  const name = (entry as Record<string, unknown>)['name'] ?? '?';
  if (isAssembly(entry)) {
    console.log(`== ${name} (assembly, skipped by harness)`);
    continue;
  }
  const kl = resolveKeylist(entry);
  console.log(`== ${name} keys=${kl.keylist.length}`);
  report('shell', buildShellFromAny(entry));
  report('inserts', buildInsertsFromAny(entry));
  if (kl.skirt) report('baseplate', buildBaseplateFromAny(entry));
  else report('walls', buildWallsFromAny(entry));
}

File: scripts/verify.py
"""Python-side twin of verify.ts: identical output format for diffing."""
import json
import sys
from collections import defaultdict

sys.path.insert(0, sys.argv[2])  # path to BlenderKeyboardGenerator/src
import core  # noqa: E402


def report(label, verts, faces):
    if not verts:
        print(f"{label}: verts=0 faces=0 tris=0 "
              f"bbox=[Infinity,Infinity,Infinity]..[-Infinity,-Infinity,-Infinity] "
              f"checksum=0.000 nonmanifold_edges=0")
        return
    minx = min(v[0] for v in verts); maxx = max(v[0] for v in verts)
    miny = min(v[1] for v in verts); maxy = max(v[1] for v in verts)
    minz = min(v[2] for v in verts); maxz = max(v[2] for v in verts)
    s = sum(x + 2 * y + 3 * z for x, y, z in verts)
    ntris = sum(max(0, len(f) - 2) for f in faces)
    edge_count = defaultdict(int)
    for f in faces:
        for i in range(len(f)):
            a, b = f[i], f[(i + 1) % len(f)]
            edge_count[frozenset((a, b))] += 1
    bad = sum(1 for c in edge_count.values() if c != 2)
    print(f"{label}: verts={len(verts)} faces={len(faces)} tris={ntris} "
          f"bbox=[{minx:.4f},{miny:.4f},{minz:.4f}]..[{maxx:.4f},{maxy:.4f},{maxz:.4f}] "
          f"checksum={s:.3f} nonmanifold_edges={bad}")


with open(sys.argv[1]) as f:
    raw = json.load(f)
entries = raw if isinstance(raw, list) else [raw]

for entry in entries:
    if core.is_assembly(entry):
        print(f"== {entry.get('name', '?')} (assembly, skipped by harness)")
        continue
    kl = core.resolve_keylist(entry)
    print(f"== {entry.get('name', '?')} keys={len(kl['keylist'])}")
    report('shell', *core.build_shell_from_any(entry))
    report('inserts', *core.build_inserts_from_any(entry))
    if kl.get('skirt'):
        report('baseplate', *core.build_baseplate_from_any(entry))
    else:
        report('walls', *core.build_walls_from_any(entry))
